c10/util/LeftRight.h

```diff
@@ -2,69 +2,165 @@
+public:
+    explicit IncrementRAII(std::atomic<int32_t> *counter): _counter(counter) {
+        ++(*_counter);
+    }
+
+    ~IncrementRAII() {
+        --(*_counter);
+    }
+private:
+    std::atomic<int32_t> *_counter;
+
+    C10_DISABLE_COPY_AND_ASSIGN(IncrementRAII);
+};
+
+}
+
 // LeftRight wait-free readers synchronization primitive
 // https://hal.archives-ouvertes.fr/hal-01207881/document
-template <typename T>
-class LeftRight {
- public:
-  LeftRight() {
-    counters_[0].store(0);
-    counters_[1].store(0);
-  }
-
-  template <typename F>
-  auto read(F&& readFunc) const -> typename std::result_of<F(const T&)>::type {
-    auto localCounterIndex = counterIndex_.load();
-    ++counters_[localCounterIndex];
-    try {
-      auto r = readFunc(data_[dataIndex_.load()]);
-      --counters_[localCounterIndex];
-      return r;
-    } catch (const std::exception& e) {
-      --counters_[localCounterIndex];
-      throw;
+template <class T>
+class LeftRight final {
+public:
+    LeftRight()
+    : _writeMutex()
+    , _foregroundCounterIndex{0}
+    , _foregroundDataIndex{0}
+    , _counters{{{0}, {0}}}
+    , _data{{{}, {}}}
+    , _inDestruction{false} {}
+
+    ~LeftRight() {
+        // from now on, no new readers/writers will be accepted (see asserts in read()/write())
+        _inDestruction = true;
+
+        // wait until any potentially running writers are finished
+        {
+            std::unique_lock<std::mutex> lock(_writeMutex);
+        }
+
+        // wait until any potentially running readers are finished
+        while (_counters[0].load() != 0 || _counters[1].load() != 0) {
+            std::this_thread::yield();
+        }
     }
-  }
-
-  // Throwing from write would result in invalid state
-  template <typename F>
-  auto write(F&& writeFunc) -> typename std::result_of<F(T&)>::type {
-    std::unique_lock<std::mutex> lock(mutex_);
-    return uniqueWrite(std::forward<F&&>(writeFunc));
-  }
-
- private:
-  // This function doesn't use any locks for the writers. Use only if you know
-  // what you're doing
-  template <typename F>
-  auto uniqueWrite(F&& writeFunc) -> typename std::result_of<F(T&)>::type {
-    try {
-      auto localDataIndex = dataIndex_.load();
-      writeFunc(data_[localDataIndex ^ 1]);
-      dataIndex_ = localDataIndex ^ 1;
-      auto localCounterIndex = counterIndex_.load();
-      while (counters_[localCounterIndex ^ 1].load()) {
-        std::this_thread::yield();
-      }
-      counterIndex_ = localCounterIndex ^ 1;
-      while (counters_[localCounterIndex].load()) {
-        std::this_thread::yield();
-      }
-      return writeFunc(data_[localDataIndex]);
-    } catch (const std::exception& e) {
-      // rethrow
-      throw;
+
+    template <typename F>
+    auto read(F&& readFunc) const -> typename std::result_of<F(const T&)>::type {
+        if(_inDestruction.load()) {
+            throw std::logic_error("Issued LeftRight::read() after the destructor started running");
+        }
```

Comment:
Can't this race with destruction? You check the condition, and then immediately after destruction on the object begins.

Reply:
Yes you're right, I have to move that check to after incrementing the counters. Done.

Btw, these destruction checks were kind of added last minute, and they're probably actually not necessary. Usually, even for threadsafe objects, it's assumed that you only start destruction once you're sure there's no thread accessing them anymore. I'll keep the checks if you don't have an opinion about it, but I'm also open to discuss removing the destruction checks.